[Eval] Evaluation of abstract causal reasoning capabilities of language model

ggendro opened the pull request:
# Thank you for contributing an eval! ♥️

🚨 Please make sure your PR follows these guidelines, __failure to follow the guidelines below will result in the PR being closed automatically__. Note that even if the criteria are met, that does not guarantee the PR will be merged nor GPT-4 access granted. 🚨

__PLEASE READ THIS__:

In order for a PR to be merged, it must fail on GPT-4. We are aware that right now, users do not have access, so you will not be able to tell if the eval fails or not. Please run your eval with GPT-3.5-Turbo, but keep in mind as we run the eval, if GPT-4 gets higher than 90% on the eval, we will likely reject since GPT-4 is already capable of completing the task.

We plan to roll out a way for users submitting evals to see the eval performance on GPT-4 soon. Stay tuned! Until then, you will not be able to see the eval performance on GPT-4. **Starting April 10, the minimum eval count is 15 samples, we hope this makes it easier to create and contribute evals.**

Also, pelase note that we're using **Git LFS** for storing the JSON files, so please make sure that you move the JSON file to Git LFS before submitting a PR. Details on how to use Git LFS are available [here](https://git-lfs.com).

## Eval details 📑
### Eval name
abstract-causal-reasoning

### Eval description

This benchmark evaluates the ability of a language to perform abstract causal reasoning using a text-based version of the Abstract Causal REasoning beyond covariation (ACRE) dataset. The task consists of finding objects causally responsible for the activation of the light in a system, from a few examples, and apply the reasoning to a test case. We provide samples in two formats, either under natural language or symbolic.

### What makes this a useful eval?

Causal structure discovery is a crucial task towards making robust and trustworthy AI systems as it aims to make a model recognise the inner mechanisms of a system from causal links and not (potentially spurious) correlations. By combining causal and abstract reasoning, this eval tests the ability of a language model to extract causal patterns from limited examples while reducing the risk of spurious correlations being used by the model to improve its performance (e.g. leak in the training data or use of other background knowledge).

## Criteria for a good eval ✅

Below are some of the criteria we look for in a good eval. In general, we are seeking cases where the model does not do a good job despite being capable of generating a good response (note that there are some things large language models cannot do, so those would not make good evals).

Your eval should be:

- [X] Thematically consistent: The eval should be thematically consistent. We'd like to see a number of prompts all demonstrating some particular failure mode. For example, we can  create an eval on cases where the model fails to reason about the physical world.
- [X] Contains failures where a human can do the task, but either GPT-4 or GPT-3.5-Turbo could not.
- [X] Includes good signal around what is the right behavior. This means either a correct answer for `Basic` evals or the `Fact` Model-graded eval, or an exhaustive rubric for evaluating answers for the `Criteria` Model-graded eval.
- [X] **Include at least 15 high quality examples.**

If there is anything else that makes your eval worth including, please document it below.

### Unique eval value

Our eval contains an extensive list of high-quality samples from a challenging and important task for making language models better reasoners.

## Eval structure 🏗️

Your eval should
- [X] Check that your data is in `evals/registry/data/{name}`
- [X] Check that your yaml is registered at `evals/registry/evals/{name}.yaml`
- [X] Ensure you have the right to use the data you submit via this eval

(For now, we will only be approving evals that use one of the existing eval classes. You may still write custom eval classes for your own cases, and we may consider merging them in the future.)

## Final checklist 👀

### Submission agreement

By contributing to Evals, you are agreeing to make your evaluation logic and data under the same MIT license as this repository. You must have adequate rights to upload any data used in an Eval. OpenAI reserves the right to use this data in future service improvements to our product. Contributions to OpenAI Evals will be subject to our usual Usage Policies (https://platform.openai.com/docs/usage-policies).

- [X] I agree that my submission will be made available under an MIT license and complies with OpenAI's usage policies.

### Email address validation

If your submission is accepted, we will be granting GPT-4 access to a limited number of contributors. Access will be given to the email address associated with the merged pull request.

- [X] I acknowledge that GPT-4 access will only be granted, if applicable, to the email address used for my merged pull request.

### Limited availability acknowledgement

We know that you might be excited to contribute to OpenAI's mission, help improve our models, and gain access to GPT-4. However, due to the requirements mentioned above and high volume of submissions, we will not be able to accept all submissions and thus not grant everyone who opens a PR GPT-4 access. We know this is disappointing, but we hope to set the right expectation before you open this PR.

- [X] I understand that opening a PR, even if it meets the requirements above, does not guarantee the PR will be merged nor GPT-4 access granted.

### Submit eval

- [X] I have filled out all required fields of this form
- [X] I have used **Git LFS** for the Eval JSON data
- [ ] (Ignore if not submitting code) I have run `pip install pre-commit; pre-commit install` and have verified that `black`, `isort`, and `autoflake` are running when I commit and push

Failure to fill out all required fields will result in the PR being closed.

### Eval JSON data 

Since we are using Git LFS, we are asking eval submitters to add in as many Eval Samples (at least 5) from their contribution here:

<details>
  <summary>View evals in JSON</summary>

  ### Eval
  ```jsonl
{"input": [{"role": "system", "content": "Objects of various color, shape, and texture are displayed. Some objects may contain a device to turn a light on if displayed. From the observations, deduce if the light is on, off, or if the state cannot be determined. Your answer must contain a single word:"}, {"role": "system", "content": "on."}, {"role": "system", "content": "off."}, {"role": "system", "content": "undetermined."}, {"role": "system", "content": "A red cube in metal is visible. The light is on."}, {"role": "system", "content": "A gray cylinder in metal is visible. The light is off."}, {"role": "system", "content": "A red cube in metal is visible. A gray cylinder in metal is visible. The light is on."}, {"role": "system", "content": "A brown cylinder in metal is visible. A yellow sphere in rubber is visible. A brown sphere in rubber is visible. A cyan cylinder in metal is visible. The light is on."}, {"role": "system", "content": "A cyan cylinder in metal is visible. The light is off."}, {"role": "system", "content": "A brown sphere in rubber is visible. A cyan cylinder in metal is visible. The light is off."}, {"role": "user", "content": "A yellow sphere in rubber is visible. The state of the light is "}], "ideal": "undetermined"}
{"input": [{"role": "system", "content": "Objects of various color, shape, and texture are displayed. Some objects may contain a device to turn a light on if displayed. From the observations, deduce if the light is on, off, or if the state cannot be determined. Your answer must contain a single word:"}, {"role": "system", "content": "on."}, {"role": "system", "content": "off."}, {"role": "system", "content": "undetermined."}, {"role": "system", "content": "A red cube in metal is visible. The light is on."}, {"role": "system", "content": "A gray cylinder in metal is visible. The light is off."}, {"role": "system", "content": "A red cube in metal is visible. A gray cylinder in metal is visible. The light is on."}, {"role": "system", "content": "A brown cylinder in metal is visible. A yellow sphere in rubber is visible. A brown sphere in rubber is visible. A cyan cylinder in metal is visible. The light is on."}, {"role": "system", "content": "A cyan cylinder in metal is visible. The light is off."}, {"role": "system", "content": "A brown sphere in rubber is visible. A cyan cylinder in metal is visible. The light is off."}, {"role": "user", "content": "A red cube in metal is visible. The state of the light is "}], "ideal": "on"}
{"input": [{"role": "system", "content": "Objects of various color, shape, and texture are displayed. Some objects may contain a device to turn a light on if displayed. From the observations, deduce if the light is on, off, or if the state cannot be determined. Your answer must contain a single word:"}, {"role": "system", "content": "on."}, {"role": "system", "content": "off."}, {"role": "system", "content": "undetermined."}, {"role": "system", "content": "A red cube in metal is visible. The light is on."}, {"role": "system", "content": "A gray cylinder in metal is visible. The light is off."}, {"role": "system", "content": "A red cube in metal is visible. A gray cylinder in metal is visible. The light is on."}, {"role": "system", "content": "A brown cylinder in metal is visible. A yellow sphere in rubber is visible. A brown sphere in rubber is visible. A cyan cylinder in metal is visible. The light is on."}, {"role": "system", "content": "A cyan cylinder in metal is visible. The light is off."}, {"role": "system", "content": "A brown sphere in rubber is visible. A cyan cylinder in metal is visible. The light is off."}, {"role": "user", "content": "A brown sphere in rubber is visible. A cyan cylinder in metal is visible. A gray cylinder in metal is visible. The state of the light is "}], "ideal": "off"}

{"input": [{"role": "system", "content": "Figure out the pattern in the following examples and apply it to the test case. Your answer must follow the format of the examples. You can answer 1 if the solution cannot be determined. Your answer must be one of the following choices:"}, {"role": "system", "content": "0."}, {"role": "system", "content": "1."}, {"role": "system", "content": "2."}, {"role": "system", "content": "[3] -> 2"}, {"role": "system", "content": "[17] -> 0"}, {"role": "system", "content": "[3, 17] -> 2"}, {"role": "system", "content": "[25, 46, 40, 29] -> 2"}, {"role": "system", "content": "[29] -> 0"}, {"role": "system", "content": "[40, 29] -> 0"}, {"role": "user", "content": "[46] -> "}], "ideal": 1}
{"input": [{"role": "system", "content": "Figure out the pattern in the following examples and apply it to the test case. Your answer must follow the format of the examples. You can answer 1 if the solution cannot be determined. Your answer must be one of the following choices:"}, {"role": "system", "content": "0."}, {"role": "system", "content": "1."}, {"role": "system", "content": "2."}, {"role": "system", "content": "[3] -> 2"}, {"role": "system", "content": "[17] -> 0"}, {"role": "system", "content": "[3, 17] -> 2"}, {"role": "system", "content": "[25, 46, 40, 29] -> 2"}, {"role": "system", "content": "[29] -> 0"}, {"role": "system", "content": "[40, 29] -> 0"}, {"role": "user", "content": "[3] -> "}], "ideal": 2}
{"input": [{"role": "system", "content": "Figure out the pattern in the following examples and apply it to the test case. Your answer must follow the format of the examples. You can answer 1 if the solution cannot be determined. Your answer must be one of the following choices:"}, {"role": "system", "content": "0."}, {"role": "system", "content": "1."}, {"role": "system", "content": "2."}, {"role": "system", "content": "[3] -> 2"}, {"role": "system", "content": "[17] -> 0"}, {"role": "system", "content": "[3, 17] -> 2"}, {"role": "system", "content": "[25, 46, 40, 29] -> 2"}, {"role": "system", "content": "[29] -> 0"}, {"role": "system", "content": "[40, 29] -> 0"}, {"role": "user", "content": "[40, 29, 17] -> "}], "ideal": 0}
  ```
</details>

ggendro commented:
Thank you for the feedback!

I proceeded to the following changes:
 1. I changed the format of the `symbolic_samples` labels from int to str
 2. I pruned the files to 2000 samples each
 
Regarding your point 3. the dataset was not created from a generator script but adapted from a visual question answering dataset (ACRE, i.e. Abstract Causal Reasoning beyond covariation) to text format.

usama-openai reviewed (changes requested):
Thanks for implementing the requested changes.

I would recommend asking the model to reason before answering and using the `Includes` evaluation method here because such complex tasks are hard for the model to do zero-shot without a chance to reason through the steps. Asking the model to provide reasoning will give the model a fair chance to solve the puzzle, and a proper evaluation method can be used to evaluate the provided answer. You also need to add instructions to provide output in a specific format, and then use that format to write the ideal answer. For example, the model can be asked to enclose the final answer in square brackets, and the ideal answer can be formatted like `[5]`. It'll give the model a proper chance to reason before answering, and the proper formatting will help in identifying the final answer properly using the `Includes` method.

ggendro commented:
Thank you for the additional feedback. I modified the evaluation from `Match `to `Includes `and added the following sentence to the pormpt: `Figure out the pattern in the following examples and apply it to the test case. Explain your reasoning step-by-step and provide the expected output immediately after writing the word ANSWER: .`
I think this method make it easier to extract the actual answer. I found that ChatGPT tends to attempt to explain each of the examples, so it's hard to disambiguate the actual answer for the test case from the answer of a previous example. Asking to write the answer after the keyword ANSWER:  solves this issue.

usama-openai reviewed (approved):
This PR looks in good shape now. I'm approving this PR.

usama-openai commented:
You should see GPT-4 API access enabled in your account in the next few days.
